vernac/canonical.ml

```diff
@@ -21,10 +21,12 @@ let cache_canonical_structure (_, (o,_)) =
   let sigma = Evd.from_env env in
   register_canonical_structure ~warn:true env sigma o
 
-let discharge_canonical_structure (_,(x, local)) =
-  if local then None else Some (x, local)
+let discharge_canonical_structure (_,((gref, _ as x), local)) =
+  if local || (Globnames.isVarRef gref && Lib.is_in_section gref) then None
```

Comment:
Why the additional check that we're in a section? Variables are not supposed to occur outside of one. Furthermore, I think this will code break for nested sections, e.g.
```Coq
Section Foo.
Variable x : X.
Section Bar.
Canonical Structure x.
End Bar.
(** ← at this point x is not canonical anymore *)
End Foo.
```

Reply:
is_in_section is what makes your nested section example work

Reply:
Aren't discharging functions called in the environment *before* closing the section? I would have sworn so...

Reply:
Oh, I just realized I misread that snippet. Forget all what I said, this is indeed correct.